packages/core/src/libraries/quota.ts

```diff
@@ -135,5 +142,124 @@ export const createQuotaLibrary = (
     }
   };
 
-  return { guardKey };
+  const guardTenantUsageByKey = async (key: keyof SubscriptionQuota) => {
+    const { isCloud, isIntegrationTest } = EnvSet.values;
+
+    // Cloud only feature, skip in non-cloud environments
+    if (!isCloud) {
+      return;
+    }
+
+    // Disable in integration tests
+    if (isIntegrationTest) {
+      return;
+    }
+
+    const { quota: fullQuota, usage: fullUsage } = await getTenantSubscriptionQuotaAndUsage(
+      cloudConnection
+    );
+
+    // Type `SubscriptionQuota` and type `SubscriptionUsage` are sharing keys, this design helps us to compare the usage with the quota limit in a easier way.
+    const { [key]: limit } = fullQuota;
+    const { [key]: usage } = fullUsage;
+
+    if (limit === null) {
+      return;
+    }
+
+    if (typeof limit === 'boolean') {
+      assertThat(
+        limit,
+        new RequestError({
+          code: 'subscription.limit_exceeded',
```

Comment:
nit: Just double-check whether we have a different error code for the bool typed quota guard